src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,467 @@
+use futures_util::future::Shared;
+use roaring::RoaringTreemap;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    num::NonZeroU64,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+    time::{Duration, Instant},
+};
+use tokio::{sync::RwLock, time::interval};
+use vector_core::event::BatchStatusReceiver;
+use warp::Rejection;
+
+use crate::shutdown::ShutdownSignal;
+use crate::sources::util::finalizer::OrderedFinalizer;
+
+use super::ApiError;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default)]
+pub struct HecAcknowledgementsConfig {
+    pub max_pending_acks: NonZeroU64,
+    pub max_number_of_ack_channels: NonZeroU64,
+    pub max_pending_acks_per_channel: NonZeroU64,
+    pub ack_idle_cleanup: bool,
+    pub max_idle_time: NonZeroU64,
+}
+
+impl Default for HecAcknowledgementsConfig {
+    fn default() -> Self {
+        Self {
+            max_pending_acks: NonZeroU64::new(10_000_000).unwrap(),
+            max_number_of_ack_channels: NonZeroU64::new(1_000_000).unwrap(),
+            max_pending_acks_per_channel: NonZeroU64::new(1_000_000).unwrap(),
+            ack_idle_cleanup: false,
+            max_idle_time: NonZeroU64::new(300).unwrap(),
+        }
+    }
+}
+
+pub struct IndexerAcknowledgement {
+    max_pending_acks: u64,
+    max_pending_acks_per_channel: u64,
+    max_number_of_ack_channels: u64,
+    channels: Arc<RwLock<HashMap<String, Arc<Channel>>>>,
```

Comment:
Would a `DashMap` make sense here instead? It has the concurrency baked in, and is intentended as a replacement for `RwLock<HashMap<…>>`. This would eliminate the locking games below.

Reply:
Ah, interesting. The docs mention that 
> Locking behaviour: May deadlock if called when holding any sort of reference into the map.
I'm not sure I understand that: is having an `Arc<Channel>` map value an issue?

Given the below race condition comment, would using a `Mutex<HashMap>>` be faster? It seems like we'll need the write lock more often than not.

Reply:
I think that deadlock issue is when calling, for example, `iter` which returns a type that holds a reference to the internals, and call `insert` while that reference is still live. I would not expect an RC type to cause any kind of problem. You may still have to be careful about query/insert race conditions, but the `entry` interface should help there.

Reply:
Oh, and to the other question, yes if we need a write lock most of the time then a mutex is likely preferred, particularly when the writing critical section is fast.

Reply:
Gotcha. Will move forward with `Mutex<HashMap>` as updated in 6365dbf